File: DuanShikai/week03-04/_155_MinStack.java
package com.sky.leetcode;


/**
 * @author DuanSky
 * @date 2015年12月16日 下午9:00:31
 * @content 
 */

public class _155_MinStack{
	
	public static void main(String args[]){
		_155_MinStack minStack=new _155_MinStack();
		minStack.push(-10);
		minStack.push(14);
		minStack.getMin();
		minStack.getMin();
		minStack.push(-20);
		minStack.getMin();
		minStack.getMin();
		minStack.top();
		minStack.getMin();
		minStack.pop();
		minStack.push(10);
		minStack.push(-7);
		minStack.getMin();
		minStack.push(-7);
		minStack.pop();
		minStack.top();
		minStack.getMin();
		minStack.pop();
	}
	
	private Node top;
	
	public void push(int x) {
        if(top==null)
        	top=new Node(x,x);
        else{
        	Node newNode=new Node(x,Math.min(getMin(),x));
        	top.forward=newNode;
        	newNode.back=top;
        	top=newNode;
        }
        
    }

    public void pop() {
        if(top!=null){
        	top=top.back;
        	if(top!=null){
        		top.forward.back=null;
        		top.forward=null;
        	}
        }
    }

    public int top() {
       return top!=null ? top.x : 0;
    }

    public int getMin() {
    	return top!=null ? top.min : 0;
    }
    
    class Node{
    	private int x;
    	private int min;
    	public Node back;
    	public Node forward;
    	Node(int x,int min){
    		this.x=x;
    		this.min=min;
    	}
    }
}

File: LuoSonglei/week-95/155-MinStack.c
/*******************************************
Author      : LHearen
E-mail      : LHearen@126.com
Time        : 2016-03-08 12:31
Description : Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.

push(x) -- Push element x onto stack.
pop() -- Removes the element on top of the stack.
top() -- Get the top element.
getMin() -- Retrieve the minimum element in the stack.
Source      : https://leetcode.com/problems/min-stack/
*******************************************/
#include <limits.h>
typedef struct
{
    int *arr;
    int count;
    int *mins;
    int minCount;
} MinStack;

void minStackCreate(MinStack *stack, int maxSize)
{
    stack->arr = (int*)malloc(sizeof(int)*maxSize);
    stack->mins = (int*)malloc(sizeof(int)*maxSize); //record the mins till the top of the arr;
    stack->count = 0;
    stack->minCount = 0;
}

void minStackPush(MinStack *stack, int element) //push it to arr normally, but meantime check whether we should push it to mins;
{
    stack->arr[stack->count++] = element;
    if(stack->minCount==0 || element<=stack->mins[stack->minCount-1])
        stack->mins[stack->minCount++] = element;
}

void minStackPop(MinStack *stack) //pop will always pop the top -> the top of mins and arr;
{
    int top = stack->arr[stack->count-1];
    if(stack->mins[stack->minCount-1] == top)
        stack->minCount--;
    stack->count--;
}

int minStackTop(MinStack *stack) //just return the top, needless to remove it;
{
    return stack->arr[stack->count-1];
}

int minStackGetMin(MinStack *stack) //just return the min, needless to remove it;
{
    return stack->mins[stack->minCount-1];
}

void minStackDestroy(MinStack *stack)
{
    free(stack->arr);
    free(stack->mins);
}
